### PR TITLE
Preserve environment variables from the parent image on dockerTools.buildImage

### PR DESCRIPTION
**Pull request by utdemir**
###### Motivation for this change

When building the Docker images with `docker build`, environment variables from the parent is inherited to the child images. This was not the case with `dockerTools.buildImage`, since it discards the parent configuration when building the child image. This behaviour is a bit inconvenient since that pattern is widely used.

This PR fetches the environment variables from the parent image and injects to the childs configuration. I also extended the NixOS tests to test this behaviour.

Still this is less than ideal, since the environment variables is not the only thing which should be inherited from the parents. Cmd, entrypoint, user, working directory, labels, volumes are also inherited, but unfortunately all in slightly different ways; and I think trying to replicate all that behavior will be costly. So, we can currently only handle environment variables and implement the others only when necessary.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by andir** [comment]
@grahamc might also be qualified to review this

**Comment by purcell** [comment]
Nice! Agree re. handling `Cmd`, `Entrypoint` etc. ultimate: this PR is definitely a worthwhile improvement in the meantime.

**Comment by bjornfor** [comment]
I looked at old PRs and found this from 2018! https://github.com/NixOS/nixpkgs/pull/37760.

**Comment by utdemir** [comment]
> I looked at old PRs and found this from 2018! #37760.

Sorry @bjornfor , I missed that. You are right, it's almost the same with this one.

I read the discussion there, and I still think merging this PR is worthwhile. I guess that was the final decision on that PR anyway, just the author has lost interest.

Also, one thing worth mentioning is that the inheritance of other variables (volumes, labels etc) is not that simple, because as far as I can see there are subtle nuances between the way they are inherited (eg. https://github.com/moby/moby/issues/5147). So, I think the proper solution would either:

* Figure out the inheritence rules per variable basis and implement them. This might take a lot of effort and significantly increase the code complexity.
* Re-use `docker` as a library (as we do with `tarsum`) to create the manifest of the child layers. I do not know if it is possible, and it'd probably still be at least as complex as the current solution.

So, in short I still think this PR is useful alone (as the original one).

**Review by nlewo — APPROVED** [review]
@GrahamcOfBorg test docker-tools

**Comment by nlewo** [comment]
Test locally passed.
